GVFS/GVFS.Upgrader/UpgradeOrchestrator.cs

```diff
@@ -0,0 +1,389 @@
+
+namespace GVFS.Upgrader
+{
+    public class UpgradeOrchestrator
+    {
+        private const EventLevel DefaultEventLevel = EventLevel.Informational;
+
+        private ProductUpgrader upgrader;
+        private ITracer tracer;
+        private InstallerPreRunChecker preRunChecker;
+        private TextWriter output;
+        private TextReader input;
+        private bool remount;
+        private bool shouldExitOnError;
+
+        public UpgradeOrchestrator(
+            ProductUpgrader upgrader, 
+            ITracer tracer,
+            InstallerPreRunChecker preRunChecker,
+            TextReader input,
+            TextWriter output,
+            bool shouldExitOnError)
+        {
+            this.upgrader = upgrader;
+            this.tracer = tracer;
+            this.preRunChecker = preRunChecker;
+            this.output = output;
+            this.input = input;
+            this.remount = false;
+            this.shouldExitOnError = shouldExitOnError;
+            this.ExitCode = ReturnCode.Success;
+        }
+
+        public UpgradeOrchestrator()
+        {
+            string logFilePath = GVFSEnlistment.GetNewGVFSLogFileName(
+                ProductUpgrader.GetLogDirectoryPath(),
+                GVFSConstants.LogFileTypes.UpgradeProcess);
+            JsonTracer jsonTracer = new JsonTracer(GVFSConstants.GVFSEtwProviderName, "UpgradeProcess");
+            jsonTracer.AddLogFileEventListener(
+                logFilePath,
+                DefaultEventLevel,
+                Keywords.Any);
+
+            this.tracer = jsonTracer;
+            this.preRunChecker = new InstallerPreRunChecker(this.tracer);
+            this.upgrader = new ProductUpgrader(ProcessHelper.GetCurrentProcessVersion(), this.tracer);
+            this.output = Console.Out;
+            this.input = Console.In;
+            this.remount = false;
+            this.shouldExitOnError = false;
+            this.ExitCode = ReturnCode.Success;
+        }
+
+        public ReturnCode ExitCode { get; private set; }
+
+        public void Execute()
+        {
+            string error = null;
+
+            if (this.upgrader.IsNoneRing())
+            {
+                this.output.WriteLine(GVFSConstants.UpgradeVerbMessages.NoneRingConsoleAlert);
+                this.output.WriteLine(GVFSConstants.UpgradeVerbMessages.SetUpgradeRingCommand);
+            }
+            else
+            {
+                try
+                {
+                    Version newVersion = null;
+                    if (!this.TryRunUpgradeInstall(out newVersion, out error))
+                    {
+                        this.ExitCode = ReturnCode.GenericError;
+                    }
+                }
+                finally
+                {
+                    string remountError = null;
+                    if (!this.TryRemountRepositories(out remountError))
+                    {
+                        remountError = Environment.NewLine + "WARNING: " + remountError;

```

Comment:
I forced a downgrade failure here, and the result is very confusing:
```
Downloading...Succeeded
Unmounting repositories...Succeeded
Installing Git version: 2.17.1.gvfs.2.5...Succeeded
Installing VFSForGit version: 1.0.18234.1...Succeeded
Mounting repositories...Failed

WARNING: GVFS error.
Mounting repo at c:\GVFSRepos\GVFS
Error: Changes to GVFS disk layout do not allow mounting after downgrade. Try mounting again using a more recent version of GVFS.
Enlistment disk layout version: 17
GVFS disk layout version: 16

The following repos failed to mount:
c:\GVFSRepos\GVFS

Press Enter to exit.
```

Problems:
1) "WARNING: GVFS Error". Is it a warning or an error?
2) On the default window size that the upgrade tool opened up, the text is wrapping, making it even harder to read:
![image](https://user-images.githubusercontent.com/8729434/45711412-57311100-bb3e-11e8-9eea-0159e4dcd009.png)
3) I see duplicate messages about the path `C:\GVFSRepos\GVFS` and it's hard to know how to make sense of it
4) There is nothing here to indicate to the user that the installation actually succeeded, even though one of the repos failed to mount. Their natural response will be to just run the installer again, which is not going to fix anything.

Reply:
This is a warning, agree with comments 2, 3 & 4. These errors are as shown in the existing manual installer (attached screenshot).

How about we display the mount error in a single line and log the detailed failure reason in the upgrade log file. Like this -

```
Downloading...Succeeded
Unmounting repositories...Succeeded
Installing Git version: 2.17.1.gvfs.2.5...Succeeded
Installing VFSForGit version: 1.0.18234.1...Succeeded
Mounting repositories...Failed

WARNING: Mounting of one or more repos failed. Refer <upgrade.log path> for more info.

Upgrade completed successfully!
Press Enter to exit.```

<img width="416" alt="screen shot 2018-09-18 at 5 18 08 pm" src="https://user-images.githubusercontent.com/378580/45717363-c1aa7700-bb67-11e8-9661-3b31d5ebe489.png">

@sanoursa @jrbriggs - Please share your thoughts.

Reply:
- Modified GVFS Service verb to log failed repos list to stderr. Upgrade tool will display stderr from GVFS CLI. 
- This new Service verb option is enabled using the command line switch "log-mount-failure-in-stderr"

<img width="518" alt="screen shot 2018-09-20 at 4 54 52 pm" src="https://user-images.githubusercontent.com/378580/45847087-cdc53e80-bcf7-11e8-8264-49ed249f04ae.png">

Reply:
### Testing suggestion 
The new command line option obviously is not available in the current release version of GVFS. After downloading and installing upgrade release, when Upgrader tool invokes newly installed GVFS CLI, the CLI wouldn't know the new command line option and reply with error (Screenshot attached.) One workaround is to use a custom GVFS build that has  TryInstallGVFSUpgrade() invocation in UpgradeOrchestrator::TryRunUpgradeInstall() method commented out.

<img width="500" alt="screen shot 2018-09-20 at 4 49 50 pm" src="https://user-images.githubusercontent.com/378580/45847472-2cd78300-bcf9-11e8-87b6-3e33454828cd.png">